include/gatekeeper_fib.h

```diff
@@ -0,0 +1,202 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _GATEKEEPER_GK_FIB_H_
+#define _GATEKEEPER_GK_FIB_H_
+
+#include <rte_ether.h>
+#include <rte_hash.h>
+
+#include "gatekeeper_net.h"
+#include "gatekeeper_lpm.h"
+
+enum gk_fib_action {
+
+	/* Forward the packet to the corresponding Grantor. */
+	GK_FWD_GRANTOR,
+
+	/*
+	 * Forward the packet to the corresponding gateway
+	 * in the front network.
+	 */
+	GK_FWD_GATEWAY_FRONT_NET,
+
+	/*
+	 * Forward the packet to the corresponding gateway
+	 * in the back network.
+	 */
+	GK_FWD_GATEWAY_BACK_NET,
+
+	/*
+	 * The destination address is a neighbor in the front network.
+	 * Forward the packet to the destination directly.
+	 */
+	GK_FWD_NEIGHBOR_FRONT_NET,
+
+	/*
+	 * The destination address is a neighbor in the back network.
+	 * Forward the packet to the destination directly.
+	 */
+	GK_FWD_NEIGHBOR_BACK_NET,
+
+	/* Drop the packet. */
+	GK_DROP,
+
+	/* Invalid forward action. */
+	GK_FIB_MAX,
+};
+
+/* The Ethernet header cache. */
+struct ether_cache {
+
+	/* Indicate whether the cache entry has been filled in or not. */
+	bool             filled;
+
+	/* Indicate whether the MAC address is stale or not. */
+	bool             stale;
+
+	/* The whole Ethernet header. */
+	struct ether_hdr eth_hdr;
+
+	/*
+	 * The count of how many times the LPM tables refer to it,
+	 * so a neighbor entry can go away only when no one referring to it.
+	 */
+	uint32_t         ref_cnt;
+};
+
+struct neighbor_hash_table {
+
+	/* The tables that store the Ethernet headers. */
+	struct ether_cache *cache_tbl;
+
+	struct rte_hash *hash_table;
+};
+
+/* The gk forward information base (fib). */
+struct gk_fib {
+
+	/* The fib action. */
+	enum gk_fib_action action;
+
+	/*
+	 * The caller that finished updating its flow table
+	 * needs to increment this counter, so that the blocks
+	 * that updating the FIB entry can finish its operation.
+	 */
+	rte_atomic16_t     num_updated_instances;
+
+	/*
+	 * The count of how many times the LPM tables refer to it,
+	 * so a fib entry can go away only when no LPM entry referring to it.
+	 */
+	uint32_t           ref_cnt;
+
+	union {
+		/*
+	 	 * The nexthop information when the action is
+		 * GK_FWD_GATEWAY_*_NET.
+	 	 */
+		struct {
+			/* The IP address of the nexthop. */
+			struct ipaddr ip_addr;
+
+			/* The cached Ethernet header. */
+			struct ether_cache *eth_cache;
+		} gateway;
+
+		struct {
+			/*
+		 	 * When the action is GK_FWD_GRANTOR, we need
+			 * the next fib entry for either the gateway or
+			 * the grantor server itself as a neighbor.
+			 */
+			struct gk_fib *next_fib;
+
+			/*
+		 	 * When the action is GK_FWD_GRANTOR, we need
+			 * the IP flow information.
+			 * The source should be the back net IP address.
+			 * The destination should be the Grantor IP address.
+		 	 */
+			struct ip_flow flow;
```

Comment:
I kept the ``struct ip_flow flow`` field to simplify the process of the IPIP encapsulation, which can directly obtain the tunnel information from the FIB entry. If we decide to get rid of the ``flow`` field, we may modify the interfaces for these functions. But it should be trivial.